lib/s3.js

```diff
@@ -0,0 +1,72 @@
+const S3 = require('aws-sdk/clients/s3')
+const s3 = new S3({ apiVersion: '2006-03-01' })
+
+/**
+ * select - select data from S3 using S3-select
+ *
+ * @param {object} params object to hold bucket name, key, input/output serialization etc.
+ * @param {object} options
+ * @return {Array<Object>}      Array of objects as selected data from s3
+ */
+async function select (params, options = {}) {
+  let buffer = Buffer.alloc(0)
+
+  return new Promise((resolve, reject) => {
+    s3.selectObjectContent(params, (err, data) => {
+      if (err) {
+        return reject(err)
+      }
+
+      if (!data) {
+        return reject(new Error('No data recieved'))
+      }
+
+      // data.Payload is a Readable Stream
+      const eventStream = data.Payload
+
+      // Read events as they are available
+      eventStream.on('data', (event) => {
+        // event.Records.Payload is a buffer containing a single record, partial records, or multiple records so concatenate buffers
+        if (event.Records) buffer = Buffer.concat([buffer, event.Records.Payload])
+      })
+
+      // Handle errors encountered during the API call
+      eventStream.on('error', (e) => {
+        return reject(e)
+      })
+
+      eventStream.on('end', () => {
+        // Convert buffer to string and then to JSON
+        const json = s3StringToJson(buffer.toString('utf8'))
+
+        // S3 can sometime append an empty line to end of file stream;
+        // Empty strings were converted to "undefined", so check last element and remove if found
+        if (json[json.length - 1] === undefined) json.pop()
+        return resolve(json)
+      })
+    })
+  })
+}
+
+/**
+ * stringToJson - convert string to Json for a collection of records
+ *
+ * @param  {String} str string containing multiple records
+ * @return {Array<Object>} data as array of objects
+ */
+function s3StringToJson (str) {
+  return str.split('\n')
+    .map(rec => {
+      try {
+        return JSON.parse(rec)
```

Comment:
think it may be a mistake to swallow this error wholesale. Can you detect the condition you're trapping with catch? At a minimum you should log an error here

Reply:
This has been reworked and encapsulated in a `try/catch`.  FWIW, failure of the JSON.parse should be rare because S3 Select does its own cleanup and error handling. 
- A file with a malformed JSON object will throw an object earlier, when the stream is acquired. 
- files that contain blank lines, have such lines removed by S3
- files that contain lines with non-JSON strings are converted to empty objects by S3